Make sure we can actually use RSTUDIO_USE_SYSTEM_DEPENDENCIES.
It requires downloads of 2 dependencies, so a new parameter is required.
See https://github.com/rstudio/rstudio/pull/19239

Author: Alexander Grund (TU Dresden)

diff --git a/src/cpp/core/CMakeLists.txt b/src/cpp/core/CMakeLists.txt
index c2248cd3c4..c16f9a501e 100644
--- a/src/cpp/core/CMakeLists.txt
+++ b/src/cpp/core/CMakeLists.txt
@@ -188,7 +188,6 @@ if (UNIX)
       find_library(UUID_LIBRARIES uuid)
       find_library(RT_LIBRARIES rt)
    endif()
-   find_package(ZLIB REQUIRED QUIET)
 
    # find apple frameworks we depend on
    if(APPLE)
@@ -201,7 +200,7 @@ if (UNIX)
       ${UTIL_LIBRARIES}
       ${UUID_LIBRARIES}
       ${RT_LIBRARIES}
-      ${ZLIB_LIBRARIES}
+      ZLIB::ZLIB
       ${CORE_SERVICES_LIBRARY})
 
    list(APPEND CORE_SYSTEM_LIBRARIES ${OPENSSL_LIBRARIES})
diff --git a/src/cpp/ext/CMakeLists.txt b/src/cpp/ext/CMakeLists.txt
index 8279f18a6a..3bea7dbf14 100644
--- a/src/cpp/ext/CMakeLists.txt
+++ b/src/cpp/ext/CMakeLists.txt
@@ -30,12 +30,13 @@ set(CMAKE_WARN_DEPRECATED FALSE)
 #
 # This function accepts the following single-value arguments:
 #
-# COMMENT:     A one-line description of the dependency, usually from the dependency itself.
-# VERSION:     The version of the dependency to be used.
-# REPOSITORY:  The (git) repository to be used when retrieving the dependency.
-# REVISION:    The (git) revision / commit hash to be used when retrieving the dependency.
-# CUSTOM:      Optional; when TRUE, the project is retrieved but not automatically added to the build.
-# PLATFORMS:   Optional; a list of platforms for which this dependency should be retrieved and built.
+# COMMENT:      A one-line description of the dependency, usually from the dependency itself.
+# VERSION:      The version of the dependency to be used.
+# REPOSITORY:   The (git) repository to be used when retrieving the dependency.
+# REVISION:     The (git) revision / commit hash to be used when retrieving the dependency.
+# CUSTOM:       Optional; when TRUE, the project is retrieved but not automatically added to the build.
+# ALWAYS_FETCH: Optional; when TRUE, it will always be downloaded instead of using a system package
+# PLATFORMS:    Optional; a list of platforms for which this dependency should be retrieved and built.
 #
 # Use the CUSTOM argument when you want to fetch an external dependency, but need to manually
 # set up a build target based on that package's sources. This is mainly useful for projects which
@@ -45,14 +46,15 @@ set(CMAKE_WARN_DEPRECATED FALSE)
 function(dependency)
 
    set(_NAME "${ARGV0}")
-   cmake_parse_arguments(PARSE_ARGV 1 "" "" "COMMENT;CUSTOM;VERSION;REPOSITORY;REVISION;CXXFLAGS" "PLATFORMS")
+   cmake_parse_arguments(PARSE_ARGV 1 "" "" "COMMENT;CUSTOM;ALWAYS_FETCH;VERSION;REPOSITORY;REVISION;CXXFLAGS" "PLATFORMS")
 
-   set(${_NAME}_VERSION    "${_VERSION}"    CACHE INTERNAL "")
-   set(${_NAME}_REPOSITORY "${_REPOSITORY}" CACHE INTERNAL "")
-   set(${_NAME}_REVISION   "${_REVISION}"   CACHE INTERNAL "")
-   set(${_NAME}_CUSTOM     "${_CUSTOM}"     CACHE INTERNAL "")
-   set(${_NAME}_PLATFORMS  "${_PLATFORMS}"  CACHE INTERNAL "")
-   set(${_NAME}_CXXFLAGS   "${_CXXFLAGS}"   CACHE INTERNAL "")
+   set(${_NAME}_VERSION      "${_VERSION}"      CACHE INTERNAL "")
+   set(${_NAME}_REPOSITORY   "${_REPOSITORY}"   CACHE INTERNAL "")
+   set(${_NAME}_REVISION     "${_REVISION}"     CACHE INTERNAL "")
+   set(${_NAME}_CUSTOM       "${_CUSTOM}"       CACHE INTERNAL "")
+   set(${_NAME}_ALWAYS_FETCH "${_ALWAYS_FETCH}" CACHE INTERNAL "")
+   set(${_NAME}_PLATFORMS    "${_PLATFORMS}"    CACHE INTERNAL "")
+   set(${_NAME}_CXXFLAGS     "${_CXXFLAGS}"     CACHE INTERNAL "")
 
    if(_PLATFORMS)
       foreach(_PLATFORM IN LISTS _PLATFORMS)
@@ -122,6 +124,7 @@ dependency(HUNSPELL
    REPOSITORY "https://github.com/hunspell/hunspell"
    REVISION   "2969be996acad84b91ab3875b1816636fe61a40e" # pragma: allowlist secret
    CUSTOM     TRUE
+   ALWAYS_FETCH TRUE
 )
 
 
@@ -141,6 +144,7 @@ dependency(WEBSOCKETPP
    REPOSITORY "https://github.com/amini-allight/websocketpp"
    REVISION   "ee8cf4257e001d939839cff5b1766a835b749cd6" # pragma: allowlist secret
    CUSTOM     TRUE
+   ALWAYS_FETCH TRUE
 )
 
 
@@ -165,7 +169,6 @@ dependency(ZLIB
    VERSION    "1.3.1"
    REPOSITORY "https://github.com/madler/zlib"
    REVISION   "51b7f2abdade71cd9bb0e7a373ef2610ec6f9daf" # pragma: allowlist secret
-   PLATFORMS  WIN32
 )
 
 function(fetch)
@@ -187,21 +190,34 @@ function(fetch)
       endif()
 
       if(${_PREFIX}_ENABLED)
-         if(RSTUDIO_USE_SYSTEM_DEPENDENCIES OR RSTUDIO_USE_SYSTEM_${_PREFIX})
-            if(CMAKE_VERSION VERSION_GREATER "3.24")
-               find_package("${_NAME}" "${${_PREFIX}_VERSION}" REQUIRED GLOBAL)
+         set(do_fetch TRUE)
+         if(DEFINED RSTUDIO_USE_SYSTEM_${_PREFIX})
+            if(RSTUDIO_USE_SYSTEM_${_PREFIX})
+               if(${_PREFIX}_ALWAYS_FETCH)
+                  message(FATAL_ERROR "RSTUDIO_USE_SYSTEM_${_PREFIX} was explicitly enabled but "
+                                      "${_NAME} must be fetched and an installed version cannot be used.")
+               endif()
+               set(do_fetch FALSE)
+            endif()
+         elseif(RSTUDIO_USE_SYSTEM_DEPENDENCIES)
+            if(${_PREFIX}_ALWAYS_FETCH)
+               message(STATUS "RSTUDIO_USE_SYSTEM_DEPENDENCIES was enabled but "
+                              "${_NAME} will still be fetched instead of using an installed version.")
             else()
-               find_package("${_NAME}" "${${_PREFIX}_VERSION}" REQUIRED)
+               set(do_fetch FALSE)
             endif()
-         else()
-            set(${_PREFIX}_FETCHED TRUE)
-
+         endif()
+         set(${_PREFIX}_FETCHED ${do_fetch})
+         set(${_PREFIX}_FETCHED ${do_fetch} PARENT_SCOPE)
+         if(do_fetch)
             FetchContent_Declare("${_NAME}"
                GIT_REPOSITORY "${${_PREFIX}_REPOSITORY}"
                GIT_TAG        "${${_PREFIX}_REVISION}"
                GIT_SHALLOW    ON
                SOURCE_SUBDIR  "${_SOURCE_SUBDIR}"
                EXCLUDE_FROM_ALL)
+         else()
+            find_package("${_NAME}" "${${_PREFIX}_VERSION}" REQUIRED GLOBAL)
          endif()
       endif()
 
@@ -224,8 +240,9 @@ function(fetch)
          FetchContent_MakeAvailable("${_NAME}")
          set(CMAKE_CXX_FLAGS "${_CMAKE_CXX_FLAGS}")
          message(STATUS "Fetching dependency ${_NAME} ${${_PREFIX}_VERSION} - success")
-         set("${_PREFIX}_SOURCE_DIR" "${CMAKE_BINARY_DIR}/_deps/${_NAME}-src" CACHE INTERNAL "")
-         set("${_PREFIX}_BINARY_DIR" "${CMAKE_BINARY_DIR}/_deps/${_NAME}-build" CACHE INTERNAL "")
+         FetchContent_GetProperties(${_NAME} SOURCE_DIR src BINARY_DIR bin)
+         set("${_PREFIX}_SOURCE_DIR" "${src}" CACHE INTERNAL "")
+         set("${_PREFIX}_BINARY_DIR" "${bin}" CACHE INTERNAL "")
       endif()
 
    endwhile()
@@ -237,16 +254,28 @@ fetch(
    fmt          FMT
    gsl-lite     GSL_LITE
    hunspell     HUNSPELL
-   rapidjson    RAPIDJSON
+   RapidJSON    RAPIDJSON
    tl-expected  TL_EXPECTED
    websocketpp  WEBSOCKETPP
    yaml-cpp     YAML_CPP
-   zlib         ZLIB)
+   ZLIB         ZLIB)
 
 # Create rapidjson target.
 add_library(rstudio-rapidjson INTERFACE EXCLUDE_FROM_ALL)
 target_compile_definitions(rstudio-rapidjson INTERFACE "-DRAPIDJSON_NO_SIZETYPEDEFINE")
-target_include_directories(rstudio-rapidjson INTERFACE "${RAPIDJSON_SOURCE_DIR}/include")
+if(TARGET RapidJSON)
+   target_link_libraries(rstudio-rapidjson INTERFACE RapidJSON)
+elseif(TARGET rapidjson)  # Alternative spelling also used in some installed packages
+   target_link_libraries(rstudio-rapidjson INTERFACE rapidjson)
+elseif(RAPIDJSON_INCLUDE_DIRS)
+   target_include_directories(rstudio-rapidjson INTERFACE "${RAPIDJSON_INCLUDE_DIRS}")
+elseif(RapidJSON_INCLUDE_DIRS)  # Alternative spelling also used in some installed packages
+   target_include_directories(rstudio-rapidjson INTERFACE "${RapidJSON_INCLUDE_DIRS}")
+elseif(RAPIDJSON_FETCHED)
+   target_include_directories(rstudio-rapidjson INTERFACE "${RAPIDJSON_SOURCE_DIR}/include")
+else()
+   message(FATAL_ERROR "RapidJSON include directory not found")
+endif()
 
 
 # Create websocketpp target.
